Add to_table() to RetrievalJob object

MattDelac opened the pull request:
**What this PR does / why we need it**:
We should add a notion of "OfflineJob" so that it will be easier to add new methods as well as accessible useful information.

For example, when you build a custom Provider, you might want to access the query coming from `pull_latest_from_table_or_query()` without executing it (as it will be executed differently)

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes #

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information about release notes, see kubernetes' guide here:
http://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
pull_latest_from_table_or_query() now returns an RetrievalJob object
Add method to_table() to RetrievalJob object
```

woop commented on sdk/python/feast/infra/offline_stores/bigquery.py:
```diff
@@ -228,6 +225,15 @@ def _infer_event_timestamp_from_dataframe(entity_df: pandas.DataFrame) -> str:
+        self.client = _get_bigquery_client(project=config.offline_store.project_id)
+
+    def to_table(self) -> pyarrow.Table:
+        return self.client.query(self.query).to_arrow()
```
Is it intentional for `BigQueryOfflineJob.to_arrow()` to return a pyarrow table? I would have expected `to_table()` to create a BQ table. Would `to_arrow_table()` or `to_arrow()` be clearer?

woop commented:
I really like this idea @MattDelac!

The one jarring thing is that we now have two types of jobs that are completely isolated. `RetrievalJob` and `OfflineJob`. Is there any reason for these to be different?

I like the idea of having a single `OfflineJob` that is configured differently through methods like `materialize` and `get_historical_features()` and context objects, and then you automatically get all the `to_sql` and other base functionality. The wrapping command (`materialize`, `get_historical_features`) can then work with the OfflineJob differently, like in one case it can use `.to_table()` or whatever if it needs to load the latest values into the online store.

What do you think about that?

MattDelac commented:
> The one jarring thing is that we now have two types of jobs that are completely isolated. `RetrievalJob` and `OfflineJob`. Is there any reason for these to be different?

Hmmm good question 🤔 

I can see if those two objects could/should be the same

MattDelac commented:
@woop I changed this PR to add a new method to the `RetrievalJob` object instead of creating a new object

woop commented:
> @woop I changed this PR to add a new method to the `RetrievalJob` object instead of creating a new object

Are you ready for a review?

MattDelac commented:
> > @woop I changed this PR to add a new method to the `RetrievalJob` object instead of creating a new object
> 
> Are you ready for a review?

Let me check this linter issue since I rebased on master

MattDelac commented:
> > @woop I changed this PR to add a new method to the `RetrievalJob` object instead of creating a new object
> 
> Are you ready for a review?

It sounds good now 👍

achals reviewed (commented):
looks good to me. Can we exercise this code path in the existing integration tests?

MattDelac commented:
> Can we exercise this code path in the existing integration tests?

Not sure I understand what you mean here ...

achals commented:
> > Can we exercise this code path in the existing integration tests?
> 
> Not sure I understand what you mean here ...

Sorry, i meant can we call `to_table` during one of the integration tests?

woop commented:
Also lgtm

achals reviewed (approved):
/lgtm

woop commented:
@achals @MattDelac I know we merged this, but I just realized that this comment was unaddressed https://github.com/feast-dev/feast/pull/1663#discussion_r659032133

How do you feel about making the method `to_arrow()`?

MattDelac commented:
> @achals @MattDelac I know we merged this, but I just realized that this comment was unaddressed [#1663 (comment)](https://github.com/feast-dev/feast/pull/1663#discussion_r659032133)
> 
> How do you feel about making the method `to_arrow()`?
> 
> Example in the wild https://googleapis.dev/python/bigquerystorage/latest/bigquery_storage_v1/library.html?highlight=arrow#google.cloud.bigquery_storage_v1.reader.ReadRowsIterable.to_arrow

Ho sorry it's my fault. Will push a quick fix where I rename it `to_arrow()` 👍
